### ISSUE TITLE
Should Onionshare be notarized for macOS?

### ISSUE THREAD

**Issue by Baccount** [micahflee/onionshare#953]
This will help give users more confidence that the software they download and run, no matter where they get it from, is not malware by showing a more streamlined Gatekeeper interface.
> With the public release of macOS 10.14.5, we require that all developers creating a Developer ID certificate for the first time notarize their apps.

https://developer.apple.com/news/?id=04102019a

**Comment by micahflee** [comment]
Good idea. Here is further documentation on how the process is actually done:
https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution

**Comment by micahflee** [comment]
I want to add notarizing the app as part of the macOS release process, but to start learning it I decided to notarize the OnionShare 2 installer. I figured this issue is a good place to start documenting what I learn.

[These docs](https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution/customizing_the_notarization_workflow#3087734) say:

> The notary service generates a ticket for the top-level file that you specify, as well as each nested file. For example, if you submit a disk image that contains a signed installer package with an app bundle inside, the notarization service generates tickets for the disk image, installer package, and app bundle.

The OnionShare release script makes a codesigned `.pkg` installer file, and inside that is a codesigned app bundle. So I assume I can just notarize the `.pkg` and it will notarize both the installer and the app bundle.

I made an app-specific password in my Apple ID account, and saved it to my login keychain under the name "onionshare-notarize".

Then I ran this to notarize it:

```
$ xcrun altool --notarize-app --primary-bundle-id "com.micahflee.onionshare" --username "micah@micahflee.com" --password "@keychain:onionshare-notarize" --file OnionShare-2.pkg 
2019-04-19 14:08:53.912 altool[27377:660911] No errors uploading 'OnionShare-2.pkg'.
RequestUUID = 42504659-e533-47cb-8bd1-87ff12dd0275
```

I can see all of my notarization jobs like this:

```
$ xcrun altool --notarization-history 0 --username "micah@micahflee.com" --password "@keychain:onionshare-notarize"

Notarization History - page 0

Date                      RequestUUID                          Status      
------------------------- ------------------------------------ ----------- 
2019-04-19 21:08:49 +0000 42504659-e533-47cb-8bd1-87ff12dd0275 in progress 

Next page value: 1555708129000
```

And I can find detailed info like this:

```
$ xcrun altool --notarization-info 42504659-e533-47cb-8bd1-87ff12dd0275 --username "micah@micahflee.com" --password "@keychain:onionshare-notarize"
2019-04-19 14:14:14.713 altool[27465:663092] No errors getting notarization info.

   RequestUUID: 42504659-e533-47cb-8bd1-87ff12dd0275
          Date: 2019-04-19 21:08:49 +0000
        Status: success
    LogFileURL: https://osxapps-ssl.itunes.apple.com/itunes-assets/Enigma113/v4/98/35/13/98351313-9ab7-bba2-bb4a-80850935d14e/developer_log.json?accessKey=1555902854_3664300332193691697_fHCZS38W9I0PV5MpeYT32LFh4PG3rvBeD59zTBsOzZq0X61ZKk%2FFMsBcnX%2BT%2FuQIyhL7cX6OCetYr3MPaE6SqdU%2B7Xieotd0TEzGJgkmOgHfKDEyqTwdQSPVDER7J9hmb1plT7iEmWE4ZgtWQjCSIVrNCmpDvtIifCi7yT8DGqg%3D
   Status Code: 0
Status Message: Package Approved
```

Oh look, it's approved. Let's try running the `--notarization-history` again:

```
$ xcrun altool --notarization-history 0 --username "micah@micahflee.com" --password "@keychain:onionshare-notarize"

Notarization History - page 0

Date                      RequestUUID                          Status  Status Code Status Message   
------------------------- ------------------------------------ ------- ----------- ---------------- 
2019-04-19 21:08:49 +0000 42504659-e533-47cb-8bd1-87ff12dd0275 success 0           Package Approved 

Next page value: 1555708129000
```

The final step is stapling the ticket to the installer. The docs say:

> Notarization produces a ticket that tells Gatekeeper that your app is notarized. After notarization completes successfully, the next time any user attempts to run your app on macOS 10.14 or later, Gatekeeper finds the ticket online. This includes users who downloaded your app before notarization.

So here's how to staple the ticket to the installer:

```
$ xcrun stapler staple OnionShare-2.pkg 
Processing: /Users/micahlee/Downloads/OnionShare-2.pkg
Processing: /Users/micahlee/Downloads/OnionShare-2.pkg
```

However, I don't think it's worth uploading a new version of `OnionShare-2.pkg`, with the ticket stapled to it. This will cause problems with the homebrew cask package, and I'll also have to make new PGP sigs. Now that I've notarized OnionShare 2, people will need internet access when they install it, and we can start the process of stapling with OnionShare 2.1.

**Comment by micahflee** [comment]
So I think the new macOS build process will be something like this:

- Build/codesign the package: `./install/build_osx.sh --release`, which generates `dist/OnionShare-[version].pkg`
- Notarize it: `xcrun altool --notarize-app --primary-bundle-id "com.micahflee.onionshare" -u "micah@micahflee.com" -p "@keychain:onionshare-notarize" --file dist/OnionShare-[version].pkg`
- Wait for it to get approved, check status with: `xcrun altool --notarization-history 0 -u "micah@micahflee.com" -p "@keychain:onionshare-notarize"`
- After it's approved, staple the ticket: `xcrun stapler staple dist/OnionShare-[version].pkg`
- PGP-sign the final, notarized and stapled, `dist/Onionshare-[version].pkg`
- Upload `pkg` and sig to github and onionshare.org

Because part of this process involves waiting for the notarization server to approve it, I think it's better to not include this in the script, but instead rather include it in documented steps for making a release.

Also, the docs suggest you include a secure timestamp with the codesigning signature. So I'll need to add `--timestamp` to the `codesign` commands in `install/build_osx.sh`.

**Comment by micahflee** [comment]
Actually, it isn't going to be that simple. I just made a built `OnionShare-2.1.dev1.pkg` and tried to notarize it, and it got rejected. Here is the reasoning:

```json
{
  "logFormatVersion": 1,
  "jobId": "735e453c-be01-4db3-be84-f89f84959624",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "OnionShare-2.1.dev1.pkg",
  "uploadDate": "2019-04-22T03:07:31Z",
  "sha256": "db594b8f0ea0461da93b6e4eb82e7e159da74830c4fb49756430d3d5469eeea7",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/MacOS/onionshare-gui",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/MacOS/onionshare-gui",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/MacOS/Tor/tor.real",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/MacOS/Tor/tor.real",
      "message": "The binary uses an SDK older than the 10.9 SDK.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/MacOS/Tor/libevent-2.1.6.dylib",
      "message": "The binary uses an SDK older than the 10.9 SDK.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/Resources/Tor/obfs4proxy",
      "message": "The signature algorithm used is too weak.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/Resources/Tor/obfs4proxy",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/Resources/Tor/obfs4proxy",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "OnionShare-2.1.dev1.pkg/com.micahflee.onionshare.pkg Contents/Payload/OnionShare.app/Contents/Resources/Tor/obfs4proxy",
      "message": "The binary uses an SDK older than the 10.9 SDK.",
      "docUrl": null,
      "architecture": "x86_64"
    }
  ]
}
```

Basically there are two issues:

OnionShare needs to be built with Enhanced Runtime Protections (#794), which has its own set of issues related to the build infrastructure, which is discussed in that issue. I think that's solvable.

OnionShare also ships with various Tor binaries that are extracted from Tor Browser, but those binaries (including `tor`, `libevent-2.1.6.dylib`, and `obfs4proxy`) have various problems: they don't have hardened runtime enabled, is compiled with an SDK older than 10.9, and uses a weak signature algorithm.

The only way to solve the second issue is if we stop extracting Tor binaries from Tor Browser and instead compile them from source as part of the OnionShare build process.

So, depending on how difficult this ends up being, this might not make it in the 2.1 release after all...

**Comment by micahflee** [comment]
Bumping this back to version 2.3

**Comment by Baccount** [comment]
If Onionshare is not notarized by January 2020, it will no longer run on macOS Catalina.  
**As a reminder, Mac software distributed outside the Mac App Store must be notarized by Apple in order to run on macOS Catalina. To make this transition easier and to protect users on macOS Catalina who continue to use older versions of software, we’ve adjusted the notarization prerequisites until January 2020.** https://developer.apple.com/news/?id=09032019a
@micahflee 
@mig5

**Comment by micahflee** [comment]
Here's some notes regarding building tor from source in macOS:

The repository is at https://gitweb.torproject.org/tor.git, and it can be git cloned from https://git.torproject.org/tor.git.

You can verify tags like `git tag -v tor-0.4.1.6`. Tags are signed by `Nick Mathewson <nickm@torproject.org>` and the Tor website links to his public key here: https://db.torproject.org/fetchkey.cgi?fingerprint=2133BC600AB133E1D826D173FE43009C4607B1FB

Alternatively, tor publishes signed source packages, also signed by Nick Mathewson. for example:

- https://dist.torproject.org/tor-0.4.1.5.tar.gz
- https://dist.torproject.org/tor-0.4.1.6.tar.gz.asc

These are the homebrew dependencies required to build:

```sh
brew install pkg-config libevent libscrypt openssl@1.1 zlib
```

Now, let's extract the source and try building:

```sh
tar -xvf tor-0.4.1.6.tar.gz
cd tor-0.4.1.6
./configure \
    --disable-dependency-tracking \
    --disable-silent-rules \
    --prefix=/usr/local \
    --with-openssl-dir=/usr/local/Cellar/openssl@1.1/1.1.1d/lib
make
```

This totally works and builds tor, but I think it's dynamically linking tor into the libraries installed via homebrew. In order to make a static tor binary (which is what we want), I think we may need to build the dependencies from source as well. Here are the relevant `./configure` options:

```
  --enable-static-openssl link against a static openssl library. Requires
                          --with-openssl-dir
  --enable-static-libevent
                          link against a static libevent library. Requires
                          --with-libevent-dir
  --enable-static-zlib    link against a static zlib library. Requires
                          --with-zlib-dir
  --enable-static-tor     create an entirely static Tor binary. Requires
                          --with-openssl-dir and --with-libevent-dir and
                          --with-zlib-dir

[...]

  --with-libevent-dir=PATH
                          specify path to libevent installation
  --with-ssl-dir=PATH     obsolete alias for --with-openssl-dir
  --with-openssl-dir=PATH specify path to openssl installation
  --with-zlib-dir=PATH    specify path to zlib installation
```

But maybe I can get away with just specifying what was installed via homebrew. Let's see...

```
./configure \
    --disable-dependency-tracking \
    --disable-silent-rules \
    --prefix=/usr/local \
    --with-libevent-dir=/usr/local/Cellar/libevent/2.1.11_1/lib \
    --with-openssl-dir=/usr/local/Cellar/openssl@1.1/1.1.1d/lib \
    --with-zlib-dir=/usr/local/Cellar/zlib/1.2.11/lib \
    --enable-static-openssl \ 
    --enable-static-libevent \ 
    --enable-static-zlib \ 
    --enable-static-tor
```

This is starting to work, but then hits this error:

```
[...]
checking for pthread.h... yes
checking for pthread_create... no
checking for pthread_condattr_setclock... no
checking whether free(NULL) works... no
configure: error: Your libc implementation doesn't allow free(NULL), as required by C99.
```

It looks like I'm hitting [this issue](https://trac.torproject.org/projects/tor/ticket/6623).

**Comment by micahflee** [comment]
It looks like this error is because my libraries aren't static (see https://trac.torproject.org/projects/tor/ticket/6623#comment:23) -- I think I'll just have to compile everything from source instead of using homebrew packages.

However, it's possible that all of this can be avoided, and that we can continue to pull binaries from Tor Browser. The link that @Baccount posted above says:

> You can now notarize Mac software that:
> 
>  * Doesn’t have the Hardened Runtime capability enabled.
>  * Has components not signed with your Developer ID.
>  * Doesn’t include a secure timestamp with your code-signing signature.
>  * Was built with an older SDK.
>  * Includes the com.apple.security.get-task-allow entitlement with the value set to any variation of true.

All of these things may now mean that I can pull binaries directly from Tor Browser and still get them notarized within OnionShare. Additionally, it looks like Tor Browser has successfully been able to notarize the current stable version, which means probably future versions will all be compatible with notarizing: https://trac.torproject.org/projects/tor/ticket/30126#comment:53

**Comment by micahflee** [comment]
Considering all of the requirements that were dropped for notarizing apps, I decided to go ahead and try notarizing the already-released `OnionShare-2.2.pkg` file... and it worked!

<img width="665" alt="Screen Shot 2019-10-15 at 10 40 29 AM" src="https://user-images.githubusercontent.com/156128/66855967-11a96800-ef39-11e9-91e1-94be96adc20c.png">

(I didn't stable the ticket to the binary for this release, but I'll do that in the future. I'm installing a new Catalina VM right now to see if I can install OnionShare 2.2 without any errors.)

So in fact, we won't need to build our own tor binaries after all. I think all I'll need to do to resolve this issue is update the release documentation.
